server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java

```diff
@@ -0,0 +1,189 @@
+ *   bits 58..63 -- zoom (0..29)
+ *   bits 29..57 -- X tile index (0..2^zoom)
+ *   bits  0..28 -- Y tile index (0..2^zoom)
+ */
+class GeoTileUtils {
+
+    /**
+     * Largest number of tiles (precision) to use.
+     * This value cannot be more than (64-5)/2 = 29, because 5 bits are used for zoom level itself (0-31)
+     * If zoom is not stored inside hash, it would be possible to use up to 32.
+     * Note that changing this value will make serialization binary-incompatible between versions.
+     * Another consideration is that index optimizes lat/lng storage, loosing some precision.
+     * E.g. hash lng=140.74779717298918D lat=45.61884022447444D == "18/233561/93659", but shown as "18/233561/93658"
+     */
+    static final int MAX_ZOOM = 29;
+
+    /**
+     * Bit position of the zoom value within hash - zoom is stored in the most significant 6 bits of a long number.
+     */
+    private static final int ZOOM_SHIFT = MAX_ZOOM * 2;
+
+    /**
+     * Bit mask to extract just the lowest 29 bits of a long
+     */
+    private static final long X_Y_VALUE_MASK = (1L << MAX_ZOOM) - 1;
+
+    /**
+     * Parse an integer precision (zoom level). The {@link ValueType#INT} allows it to be a number or a string.
+     *
+     * The precision is expressed as a zoom level between 0 and {@link #MAX_ZOOM} (inclusive).
+     *
+     * @param parser {@link XContentParser} to parse the value from
+     * @return int representing precision
+     */
+    static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
+        final Object node = parser.currentToken().equals(XContentParser.Token.VALUE_NUMBER)
+            ? Integer.valueOf(parser.intValue())
+            : parser.text();
+        return XContentMapValues.nodeIntegerValue(node);
+    }
+
+    /**
+     * Assert the precision value is within the allowed range, and return it if ok, or throw.
+     */
+    static int checkPrecisionRange(int precision) {
+        if (precision < 0 || precision > MAX_ZOOM) {
+            throw new IllegalArgumentException("Invalid geotile_grid precision of " +
+                precision + ". Must be between 0 and " + MAX_ZOOM + ".");
+        }
+        return precision;
+    }
+
+    /**
+     * Encode lon/lat to the geotile based long format.
+     * The resulting hash contains interleaved tile X and Y coordinates.
+     * The precision itself is also encoded as a few high bits.
+     */
+    static long longEncode(double longitude, double latitude, int precision) {
+        // Mathematics for this code was adapted from https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Java
+
+        // Number of tiles for the current zoom level along the X and Y axis
+        final int tiles = 1 << checkPrecisionRange(precision);
+
+        int xTile = (int) Math.floor((normalizeLon(longitude) + 180) / 360 * tiles);
+
+        double latSin = Math.sin(Math.toRadians(normalizeLat(latitude)));
```

Comment:
are we concerned that the latitude is not clipped here to ensure that the projection is square?

```
latitude = Math.min(Math.max(lat,  -85.05112878),  85.05112878);
```

this is regarding:

> The latitude and longitude are assumed to be on the WGS 84 datum. Even though Bing Maps uses a spherical projection, it’s important to convert all geographic coordinates into a common datum, and WGS 84 was chosen to be that datum. The longitude is assumed to range from -180 to +180 degrees, and the latitude must be clipped to range from -85.05112878 to 85.05112878. This avoids a singularity at the poles, and it causes the projected map to be square. [1]

and

> Because the Mercator projects the poles at infinity, Google Maps cannot show the poles. Instead it cuts off coverage at 85.051129° north and south. This is not considered a limitation, given the purpose of the service. The value 85.051129° is the latitude at which the full map becomes a square...[2]

[1] https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system
[2] https://en.wikipedia.org/wiki/Web_Mercator_projection

Reply:
also, should there be tests for points in this region?

Reply:
We already test for points in that region, e.g.  `assertEquals(0x77FFFF4580000000L, longEncode(179.999, 89.999, 29));`, but i will add a few more. The mathematics automatically takes care of those values, assigning geopoints in that region with the min/max Y value.